Modify llvm-dwp to be able to emit string tables over 4GB without losing data

clayborg opened the pull request:
We can change llvm-dwp to emit DWARF64 version of the .debug_str_offsets tables for .dwo files in a .dwp file. This allows the string table to exceed 4GB without truncating string offsets into the .debug_str section and losing data. llvm-dwp will append all strings to the .debug_str section for a .dwo file, and if any of the new string offsets exceed UINT32_MAX, it will upgrade the .debug_str_offsets table to a DWARF64 header and then each string offset in that table can now have a 64 bit offset.

Fixed LLDB to be able to successfully load the 64 bit string tables in .dwp files.

Fixed llvm-dwarfdump and LLVM DWARF parsing code to do the right thing with DWARF64 string table headers.

clayborg committed:
Modify llvm-dwp to be able to emit string tables over 4GB without losing data.

We can change llvm-dwp to emit DWARF64 version of the .debug_str_offsets tables for .dwo files in a .dwp file. This allows the string table to exceed 4GB without truncating string offsets into the .debug_str section and losing data. llvm-dwp will append all strings to the .debug_str section for a .dwo file, and if any of the new string offsets exceed UINT32_MAX, it will upgrade the .debug_str_offsets table to a DWARF64 header and then each string offset in that table can now have a 64 bit offset.

dwblaikie commented:
Yep, sounds about right to me as an implementation of the suggestion that I think @probinson first made when we discussed this upstream (& when I went down the Simplified Template Names direction) - this only uses DWARF64 .debug_str_offsets for the specific contrtibutions that need it (so should have (worth testing, etc) no impact/bit identical output for any DWP that's already correct).

The patches will need to be separated - ideally consumers are fixed before producers (but some producers don't have good ways of testing without the producer - in which case there might be a small window of breakage (well, breakage where an existing user would've got an error on dwp output, doesn't get an error but then some tool might not be able to consume it)).

Probably needs a flag for dwp (I haven't checked whether the patch proposes one) since not all consumers would be ready for this behavior (ie: gdb, presumably).

Implementation-wise, I think there's a few utilities we have for handling the DWARF32/64 length parsing and emission that could be used in places.

clayborg commented:
Here is the separate patch for the LLDB DWARF parser with a test:

https://github.com/llvm/llvm-project/pull/167997

clayborg committed:
Merge with upstream and remove extra code that isn't needed.

clayborg commented:
I merged with upstream to get the changes from the two PRs that add DWARF64 .debug_str_offsets support for .dwo files in LLDB and LLVM.

clayborg commented:
I was thinking I could also add an option, possibly keeping it hidden by default, so that in testing I can force promotion from DWARF32 to DWARF64 so we can test that it works. 

Looking at the way this was coded, it seems like llvm-dwp would have missed the fact that the `.debug_str` went over 4GB limit when writing the strings and would have truncated the new offset. There is some code that uses a variable for `OverflowOptValue` to try and catch issues or ignore them, but everything I see seems like it is 32 bit offsets so I am not sure how it would catch this.

I will add an option so I can test without creating a huge .dwp file and I will report back on what I find about how LLDB and GDB deal with such input, and if they silently fail, then I will add an option to enable this.

clayborg committed:
Add new --force-dwarf64-str-offsets option to llvm-dwp and a test that uses it.

This patch adds a new llvm-dwp option that can be used in testing to verify that llvm-dwp can successfully upgrade a .debug_str_offsets tables from DWARF32 to DWARF64.

clayborg committed:
Add a --dwarf64-str-offsets option value.

This option controls if llvm-dwp can promote a .debug_str_offsets table from DWARF32 to DWARF64.

Setting this option value to "enabled" allows promotion of a DWARF32 .debug_str_offsets table to DWARF64 only if any string in the .debug_str_offsets table exceeds UINT32_MAX.

Setting this option value to "disabled" (the default) will keep pre-existing behavior where all .debug_str_offsets tables will be emitted in the same format as are in each .dwo file.

Setting this option value to "always" forces all .debug_str_offsets tables to be emitted as DWARF64 tables. This is used for testing.

Removed the previous --force-dwarf64-str-offsets option.

clayborg committed:
Fix a typo.

clayborg committed:
Clang format.

clayborg committed:
Fix help text to represent the right default value if --dwarf64-str-offsets is specified without a value.

clayborg committed:
More cleanup on help text for --dwarf64-str-offsets option.

clayborg committed:
Fix llvm-dwp assertion errors.

I was creating a llvm-dwp file that had a .debug_str section that
exceeded 4GB and some asserts were firing.

grigorypas commented on llvm/include/llvm/DWP/DWP.h:
```diff
@@ -22,6 +22,12 @@ enum OnCuIndexOverflow {
   Continue,
 };
 
+enum Dwarf64StrOffsets {
```
Maybe Dwarf64StrOffsetsPromotion or similar? To make it clear that we are not downgrading to 32bit existing 64bit offsets.

clayborg committed:
Rename some variables and options per review suggestions.

Renamed --dwarf64-str-offsets to --dwarf64-str-offsets-promotion
Renamed Dwarf64StrOffsets type to Dwarf64StrOffsetsPromotion.

dwblaikie commented on llvm/lib/DWP/DWP.cpp:
```diff
@@ -413,33 +413,52 @@ Expected<InfoSectionUnitHeader> parseInfoSectionUnitHeader(StringRef Info) {
+    OffsetRemapping[PrevOffset] = NewOffset;
+    // Only promote the .debug_str_offsets to DWARF64 if our setting allows it.
+    if (StrOffsetsOptValue != Dwarf64StrOffsetsPromotion::Disabled &&
+        NewOffset > UINT32_MAX) {
```
We should probably have an error message here if 32->64 promotion is disabled and the offset is too large for 32 bits. But that can/should be in a separate patch (since we didn't have a diagnostic for it before, this patch isn't regressing anything)

dwblaikie commented on llvm/lib/DWP/DWP.cpp:
```diff
@@ -461,16 +480,52 @@ void writeStringsAndOffsets(MCStreamer &Out, DWPStringPool &Strings,
-      writeNewOffsetsTo(Out, Data, OffsetRemapping, Offset, ContributionEnd);
+
+      StringRef HeaderBytes = Data.getBytes(&Offset, HeaderSize);
+      if (OldOffsetSize == 4 && NewOffsetSize == 8) {
```
Perhaps we could simpilfy this code by just emitting whatever the new offset section header is - there's probably not much perrformance to be saved by copying the old header in rather than emitting a new one.

dwblaikie commented on llvm/include/llvm/DWP/DWP.h:
```diff
@@ -22,6 +22,12 @@ enum OnCuIndexOverflow {
   Continue,
 };
 
+enum Dwarf64StrOffsetsPromotion {
```
Could be an enum class

dwblaikie reviewed (commented):
(probably could've done with a few more eyes on final sign-off, but don't think it's worth reverting either)
